clients/drcachesim/drpt2trace/pt2ir.cpp

```diff
@@ -0,0 +1,395 @@
+                break;
+            dx_decoding_error(status, "sync error", insn.ip);
+            return PT2IR_CONV_SYNC_PACKET_ERROR;
+        }
+
+        /* Decode the raw trace data surround by PSB. */
+        for (;;) {
+            int nextstatus = status;
+            int errcode = 0;
+            struct pt_image *image;
+
+            /* Before starting to decode instructions, we need to handle the event before
+             * the instruction trace. For example, if a mmap2 event happens, we need to
+             * switch the cached image. */
+            while (nextstatus & pts_event_pending) {
+                struct pt_event event;
+
+                nextstatus = pt_insn_event(pt_instr_decoder_, &event, sizeof(event));
+                if (nextstatus < 0) {
+                    errcode = nextstatus;
+                    dx_decoding_error(errcode, "get pending event error", insn.ip);
+                    return PT2IR_CONV_GET_PENDING_EVENT_ERROR;
+                }
+
+                /* Use a sideband session to check if pt_event is an image switch event.
+                 * If so, change the image in 'pt_instr_decoder_' to the target image. */
+                image = NULL;
+                errcode =
+                    pt_sb_event(pt_sb_session_, &image, &event, sizeof(event), stdout, 0);
+                if (errcode < 0) {
+                    dx_decoding_error(errcode, "handle sideband event error", insn.ip);
+                    return PT2IR_CONV_HANDLE_SIDEBAND_EVENT_ERROR;
+                }
+
+                /* If it is not an image switch event, the PT instruction decoder
+                 * will not switch their cached image.*/
+                if (image == NULL)
+                    continue;
+
+                errcode = pt_insn_set_image(pt_instr_decoder_, image);
+                if (errcode < 0) {
+                    dx_decoding_error(errcode, "set image error", insn.ip);
+                    return PT2IR_CONV_SET_IMAGE_ERROR;
+                }
+            }
+            if ((nextstatus & pts_eos) != 0) {
+                break;
+            }
+
+            /* Decode PT raw trace to pt_insn. */
+            status = pt_insn_next(pt_instr_decoder_, &insn, sizeof(insn));
+            if (status < 0) {
+                dx_decoding_error(status, "get next instruction error", insn.ip);
+                return PT2IR_CONV_DECODE_NEXT_INSTR_ERROR;
+            }
+
+            pt_instr_list_.push_back(insn);
+
+            /* TODO i#5505: Use drdecode to decode insn(pt_insn) to inst_t. */
+        }
+    }
+    return PT2IR_CONV_SUCCESS;
+}
+
+uint64_t
+pt2ir_t::get_instr_count()
+{
+    return pt_instr_list_.size();
+}
+
+void
+pt2ir_t::print_instrs_to_stdout()
+{
+    void *dcontext = GLOBAL_DCONTEXT;
+    for (auto it = pt_instr_list_.begin(); it != pt_instr_list_.end(); it++) {
+        disassemble_with_info(dcontext, it->raw, STDOUT, false, false);
+    }
+}
+
+bool
+pt2ir_t::init()
```

Comment:
I think it would be easier to read to move this up below the constructor.

Reply:
Done.